resolwe/flow/execution_engines/workflow/__init__.py

```diff
@@ -105,7 +105,11 @@ def evaluate(self, data):
                 contributor=data.contributor,
                 input=data_input,
             )
-            data_object.parents.add(data)
+            DataDependency.objects.create(
```

Comment:
I don't like this logic outside of model. Can you move it to the model method, e.g. `Data.add_subprocess_dependency(<parent>)`.

Reply:
This is the Django-documented standard way to do this, why create another method? I don't like your approach as it requires a new method for each type.

Reply:
You can pass the type to the method, it is OK for me. The main thing I don't like ist that intermediate model, which should be transparent (in Django's many to many relations), is not transparent anymore.

Reply:
It is transparent in queries, just not in `add`, `create` or `set` as mentioned in [Django docs](https://docs.djangoproject.com/en/1.10/topics/db/models/#extra-fields-on-many-to-many-relationships). I can add this easily, but I think it is inconsistent about how this is handled elsewhere, e.g. `PositionInRelation` is one such intermediate model, which is always referenced directly and there are no similar helpers.

Note that we will most probably want to expose the dependency between data instances via our API as well (e.g., to show computation graphs in the frontend), which means that `DataDependency` may get its own endpoint.

Reply:
Ok, then leave it this way. I didn't know there is a plan to publicly expose this.